patterns/README.md

```diff
@@ -1030,3 +1030,371 @@ A field MAY have a `missingValues` property that MUST be an `array` where each e
+   rather than replacing them. This is so that software can distinguish
+   between so-called *system missing values* (e.g., "Not applicable") and
+   other values that you may wish to include in certain tabulations/analyses
+   but exclude from others (e.g., "Don't know" or "Refused").
+2. Add an optional field-specific `ordered` property, which can be used when
+   contructing a categorical (or factor) to indicate that the variable is
+   ordinal.
+3. Add an optional field-specific `encoding` property for use when data are
+   stored using integer or other codes rather than using the category labels.
+   This contains an object mapping the codes appearing in the data (keys) to
+   what they mean (values), and can be used by software to construct
+   corresponding value labels or categoricals (when supported) or to translate
+   the values when reading the data.
+
+As none of the three proposed properties is part of the current
+[table schema](https://specs.frictionlessdata.io//table-schema/), the proposed
+extensions are fully backward compatible.
+
+Here is an example using extensions (1) and (2):
+
+```
+{
+  "fields": [
+    {
+      "name": "physical_health",
+      "type": "string",
+      "constraints": {
+        "enum": [
+          "Poor",
+          "Fair",
+          "Good",
+          "Very good",
+          "Excellent",
+        ]
+      }
+      "ordered": true
+      "missingValues": ["Don't know","Refused"]
+    }
+  ],
+  "missingValues": ["Not applicable","No answer"]
+}
+```
+
+This is our preferred strategy, as it provides all of the information
+necessary to support fully the categorical functionality of the software
+packages listed above, while still yielding a useable result for software
+without such capability. As described below, value labels or categoricals can
+be created automatically based on the ordering of the values in the `enum`
+array, and the field level `missingValues` can be incorporated into the value
+labels or categoricals if desired. In those cases where it is desired to have
+more control over how the value labels are constructed, this information can
+be stored in a separate encodings file in JSON format or as part of a custom
+extension to the table schema. Since such instructions do not describe the
+data themselves (but only how a specific software package should handle them),
+and since they are often software- and/or user-specific, we argue that they
+should not be included in the official table schema.
+
+Alternatively, those who wish to store their data in encoded form (e.g., this
+is the default for data exports from [REDCap](https://projectredcap.org), a
+commonly-used platform for collecting data for clinical studies) may use
+extension (3) to do so:
+
+```
+{
+  "fields": [
+    {
+      "name": "physical_health",
+      "type": "integer",
+      "enum": [1,2,3,4,5]
+      "ordered": true
+      "missingValues": ["Don't know","Refused"]
```

Comment:
I assume the field `physical_health` in this example could then contain values such as:

```
1
5
Don't know
No answer
```

And because `Don't know` and `No answer` are defined as missingValues, it meets the requirements for `integer`?

Reply:
That's exactly correct; those values would be considered valid just as they are now (i.e., under the current schema). And software could handle them in 4 ways (presumably controlled via one or more options):

1. Drop the "Don't know" and "No answer" responses, and store the resulting variable as an integer (this would presumably be available for all software)
2. Create a value label mapping "Don't know" and "No answer" to extended missing values (Stata or SAS) or negative integers (SPSS), store the resulting variable as an integer and apply the value label
3. Create a categorical or factor variable, optionally including "Don't know" and "No answer" as possible categories (Python, R or Julia)
4. Translate the variable to string, and retain "Don't know" and "No answer" as valid values (again, presumably available for all software)

Reply:
> those values would be considered valid just as they are now (i.e., under the current schema).

But that would only be the case if the current (or suggested) **pattern** [missing values per field](https://specs.frictionlessdata.io/patterns/#missing-values-per-field) is implemented, correct? I am not aware of much software that handles this pattern, so it would result in `No answer` (part of the schema's `missingValues`) being converted to a missing value (passing validation), while `Don't know` (part of the field's `missingValues`) is not converted and would throw a data type error.

**I think the pattern shouldn't introduce validation errors if it is not implemented.** So I would suggest that field values should always be of the correct type, irrespective of the field's `missingValues` (i.e. if you want to use `Don't know`, the field should be a string). Software can then build upon that to convert to factors/integers.

---

> Drop the "Don't know" and "No answer" responses, and store the resulting variable as an integer (this would presumably be available for all software) ...

I think the **default implementation** (if adopted by software) should be the same as for `missingValues` (https://specs.frictionlessdata.io/table-schema/#missing-values and https://specs.frictionlessdata.io/patterns/#missing-values-per-field):

> `missingValues` dictates which string values should be treated as `null` values. This conversion to `null` is done before any other attempted type-specific string conversion.
> 
> Values that match the `missingValues` are treated as `null`.

Reply:
This is an important point, and one I must confess I hadn't fully considered. I was loosely thinking of field-specific `missingValues` as functioning similarly to schema-level `missingValues`, meaning that values specified would be converted to `null` before type-specific string conversion. In that case, a field could be designated as `type` integer even if it contains values such as "Don't know" that are included in that field's `missingValues`. But as @peterdesmet correctly points out, that means that data files relying on this would fail validation with software that doesn't implement field-specific `missingValues`.

I am persuaded by the argument that we don't want to be creating data files that are considered invalid by software that does not implement field-specific `missingValues`. At the same time, though, I would just point out that that limits what can be done with field-specific `missingValues`—they can be used to provide additional information to software that implements them, but they cannot be used in place of schema-level `missingValues`, if, say, you wanted to indicate different missing values for different fields and have them converted to `null` prior to type-specific string conversion. In a nutshell, I believe the tradeoff is between backward compatibility with existing software and having a field-specific `missingValues` property that functions the same as the existing schema-level property.

The original objective of this pattern was to fully support the use of value labels and categoricals, and the change that @peterdesmet suggests, namely requiring `type` string in cases where field-specific `missingValues` contain non-numeric values, IMO doesn't limit the ability of software to do that. Thus, I'll wait a bit to see if anyone else wants to weigh in on this, and if not I'll go ahead and make the suggested change.

Reply:
Ok, my apologies for belaboring this, but as I started to make the change @peterdesmet suggested above, I realized that aspects of what I have proposed here may ultimately clash with other proposals for field-specific `missingValues`. Specifically:

1. The intended purpose of field-specific `missingValues` within the context of this proposal was to permit software to automatically distinguish between so-called _system missing values_ (e.g., "", "NA", "."), which are always excluded, and those missing values that are sometimes retained in tabulations and certain statistical models (e.g., "Don't know", "Refused", etc.). The latter may be represented with special constructs (e.g., _extended missing values_) in Stata, SAS and SPSS, and may be included when creating categoricals/factors, if desired. My thought was to permit use of the resource level `missingValues` for system missing, and the field-specific `missingValues` for extended missing. Software that does not implement field-specific `missingValues` could just retain these values in the data, which seemed like a reasonable default.
2. In re-reviewing the previous discussions on field-specific `missingValues` (e.g., #551 and [here](https://specs.frictionlessdata.io/patterns/#missing-values-per-field)), I see that the consensus here is that when these are both present, the more specific one (i.e., the field-specific `missingValues`) should _replace_ the less specific one (i.e., the resource level `missingValues`)—not be combined with it as I had originally proposed. (I presume that when folks are referring to [cascading](https://github.com/frictionlessdata/specs/issues/551#issuecomment-354780949) here, they are referring to the way CSS works, i.e., the property provided at the most specific level takes precedence.) I understand the logic for that (e.g., more explicit, easier to implement). However, if that were implemented in software, it would yield a sub-optimal result in the case I just described in (1), i.e., the extended missing values would be dropped but the system missing values would be retained! This would violate the principle we articulated for this proposal of yielding a reasonable result for all software (including existing).

For this reason, I am now inclined to remove the field-specific `missingValues` property entirely from this proposal. That addresses @peterdesmet's concern, and will not create any problems if field-specific `missingValues` are ultimately implemented according to what appears to be the current consensus. In addition, it creates only a minor limitation for software that wants to implement value labels or categoricals—namely, there won't be any basis for guessing at what should be considered system missing and what should be considered extended missing. Instead, such software will have to provide the user with options for recognizing system missing, which should work well in most cases.

I'll give it one more week before making this proposed change, just in case anyone wants to weigh in (@khughitt, @peterdesmet, @ipimpat).

Reply:
Thanks for summarizing. I'm happy with your suggestion to not implement field specific `missingValues` as part of this proposal. The other aspects of this proposal are really useful on their own and I'm looking forward to see them adopted.

Reply:
Ok, I've removed the field specific `missingValues` property from this pattern as discussed above, though I've left a note explaining the rationale in case this issue comes up again in the future.